File: app/presenters/recommendation_presenter.rb
class RecommendationPresenter
  # Language
  LANGUAGE_SCORE_LIMIT = 45.0
  FULL_LANGUAGE_SCORE = 45.0
  # Primary Tag
  PRIMARY_TAG_SCORE_LIMIT = 15.0
  FULL_PRIMARY_TAG_SCORE = 15.0
  # Rounded up to 8 instead of 7.5
  HALF_PRIMARY_TAG_SCORE = 8.0
  # Search tokens
  SEARCH_TOKEN_SCORE_LIMIT = 40.0
  # Total
  # Language + Primary Tag + Search Tokens
  TOTAL_SCORE_LIMIT = 100.0
  # Total without language
  # Primary Tag + Search Tokens / Total
  SCORE_LIMIT_WITHOUT_LANGUAGE = 0.55
  # Total without language and primary tag
  # Search Tokens / Total
  SCORE_LIMIT_WITHOUT_LANGUAGE_AND_PRIMARY_TAG = 0.4
  RECOMMENDATION_COUNT = 4

  attr_reader :object, :options, :cid, :slug_id, :language_cid, :primary_tag,
              :search_tokens_array, :is_listed, :search_tokens_size,
              :recommendation_rankings, :min_rankings

  # Cache the recommendable snippets on the class to avoid multiple queries.
  @@recommendable_snippets = nil
  @@recommendable_snippets_by_language = {}
  @@recommendable_snippets_by_language_and_primary_tag = {}

  # Ugly but efficient optimization strategy to first check the most relevant snippets.
  def self.prepare_recommendable_snippets
    @@recommendable_snippets = Snippet.preload(:language).published.ranked.group_by(&:language_cid)

    @@recommendable_snippets.each do |language_cid, snippets|
      @@recommendable_snippets_by_language[language_cid] =
        snippets + @@recommendable_snippets.except(language_cid).values.flatten

      @@recommendable_snippets_by_language_and_primary_tag[language_cid] = {}

      tag_grouped_snippets = snippets.group_by { |s| s.primary_tag }
      tag_grouped_snippets.each do |primary_tag, primary_tag_snippets|
        @@recommendable_snippets_by_language_and_primary_tag[language_cid][primary_tag] =
          primary_tag_snippets +
          tag_grouped_snippets.except(primary_tag).values.flatten +
          @@recommendable_snippets.except(language_cid).values.flatten
      end
    end

    @@recommendable_snippets = @@recommendable_snippets.values.flatten
  end

  def self.recommendable_snippets(language = nil, tag = nil)
    prepare_recommendable_snippets if @@recommendable_snippets.blank?

    return @@recommendable_snippets if language.blank?

    return @@recommendable_snippets_by_language[language] unless tag.present? &&
      @@recommendable_snippets_by_language_and_primary_tag[language].key?(tag)

    @@recommendable_snippets_by_language_and_primary_tag[language][tag]
  end

  def initialize(object, options: {})
    @object = object
    @options = options

    @cid = object.cid
    @slug_id = object.slug_id
    @language_cid = object.language_cid
    @primary_tag = object.primary_tag
    @search_tokens_array = object.search_tokens_array
    @is_listed = object.is_listed?
    @search_tokens_size = @search_tokens_array.size
    @recommendation_rankings = {}
    @min_rankings = []
  end

  def recommend_snippets
    recommendable_snippets = RecommendationPresenter.recommendable_snippets(language_cid, primary_tag)

    recommendable_snippets.each do |snippet|
      # Skip if the snippet is the same as the current snippet
      next if snippet.cid == cid
      # Skip if the snippet is the same in another language
      next if snippet.slug_id == slug_id
      # Skip unless this snippet is listed or the object is unlisted
      next unless !is_listed || snippet.is_listed?

      # Store the minimum ranking so far
      min_ranking = min_rankings.first || 0.0

      # Performance optimization - if language score is 0 and the minimum
      # recommendation score is greater than the score limit without language,
      # then we can skip the rest of the calculations.
      is_same_language = snippet.language_cid == language_cid
      next if !is_same_language &&
        min_ranking > SCORE_LIMIT_WITHOUT_LANGUAGE

      # Determine score for language:
      #  * Same language, as language = 100% of language score
      #  * Not same language = 0% of language score
      language_score = is_same_language ? FULL_LANGUAGE_SCORE : 0

      # Performance optimization - if both language and primary tag scores are
      # 0 and the minimum recommendation score is greater than the score limit
      # without language and primary tag, then we can skip the rest of the
      # calculations.
      primary_tag_index =
        is_same_language ? snippet.tags.index(primary_tag) : nil

      next if !is_same_language &&
        primary_tag_index.blank? &&
        min_ranking > SCORE_LIMIT_WITHOUT_LANGUAGE_AND_PRIMARY_TAG

      # Determine primary tag score:
      #  * Different language = 0% of tag score
      #  * Same primary tag = 100% of tag score
      #  * Contains primary tag, but not primary = 50% of tag score
      #  * Doesn't contain tag = 0% of language score
      primary_tag_score =
        if primary_tag_index.blank?
          0
        else
          primary_tag_index == 0 ? FULL_PRIMARY_TAG_SCORE :
                      HALF_PRIMARY_TAG_SCORE
        end

      # Determine search token score:
      #  * Count found tokens and divide by total number of tokens
      matched_tokens_count = (search_tokens_array & snippet.search_tokens_array).size
      search_token_score =
        (matched_tokens_count.to_f / search_tokens_size) * SEARCH_TOKEN_SCORE_LIMIT

      # Divide by the limit to get a value between 0 and 1
      recommendation_ranking =
        (language_score + primary_tag_score + search_token_score).to_f /
        TOTAL_SCORE_LIMIT

      # Performance optimization to minimize the number of times we have to
      # sort afterwards. As soon as the minimum amount of snippets has been
      # considered, we can start trimming off any snippets below the lowest
      # snippet's recommendation ranking.
      if recommendation_ranking > 0
        if min_rankings.size < RECOMMENDATION_COUNT
          # First 4 snippets are always added
          min_rankings << recommendation_ranking
        else
          # If the new ranking is lower than the lowest ranking, ignore it
          next if recommendation_ranking < min_rankings.first

          # Otherwise, replace the lowest ranking with the new ranking
          min_rankings[0] = recommendation_ranking
        end
        min_rankings.sort!

        recommendation_rankings[snippet.id] = [
          recommendation_ranking, snippet.ranking, snippet
        ]
      end
    end

    recommendation_rankings.
      sort do |a, b|
        a[1] <=> b[1] || a[2] <=> b[2]
      end.
      reverse.
      first(RECOMMENDATION_COUNT).
      map { |s| s.last.last }
  end
end

File: src/presenters/recommendationPresenter.js
import Snippet from '#src/models/snippet.js';
import settings from '#src/config/settings.js';

export default class RecommendationPresenter {
  static candidates = null;
  static candidatesByLang = {};
  static candidatesByLangAndTag = {};

  static prepareRecommendableSnippets() {
    this.candidates = Snippet.scope('published', 'byRanking');

    const groupedCandidates = this.candidates.reduce((acc, snippet) => {
      if (!acc[snippet.languageId]) acc[snippet.languageId] = [];
      acc[snippet.languageId].push(snippet);
      return acc;
    }, {});

    Object.entries(groupedCandidates).forEach(
      ([language, languageSnippets]) => {
        const candidatesExceptLanguage = this.candidates.filter(
          snippet => snippet.languageId !== language
        );
        this.candidatesByLang[language] = languageSnippets.concat(
          candidatesExceptLanguage
        );

        const tagGroupedSnippets = languageSnippets.reduce((acc, snippet) => {
          if (!acc[snippet.primaryTag]) acc[snippet.primaryTag] = [];
          acc[snippet.primaryTag].push(snippet);
          return acc;
        }, {});

        this.candidatesByLangAndTag[language] = {};

        Object.entries(tagGroupedSnippets).forEach(([tag, tagSnippets]) => {
          const languageSnippetsExceptTag = languageSnippets.filter(
            snippet => snippet.primaryTag !== tag
          );
          this.candidatesByLangAndTag[language][tag] = tagSnippets
            .concat(languageSnippetsExceptTag)
            .concat(candidatesExceptLanguage);
        });
      }
    );
  }

  recommendableSnippets(language = null, tag = null) {
    if (RecommendationPresenter.candidates === null)
      RecommendationPresenter.prepareRecommendableSnippets();

    if (language === null) return RecommendationPresenter.candidates;

    const tagExists = Boolean(
      RecommendationPresenter.candidatesByLangAndTag[language][tag]
    );

    if (tag === null || !tagExists)
      return RecommendationPresenter.candidatesByLang[language];

    return RecommendationPresenter.candidatesByLangAndTag[language][tag];
  }

  constructor(object, options = {}) {
    this.object = object;
    this.options = options;

    this.id = this.object.id;
    this.slugId = this.object.slugId;
    this.languageId = this.object.languageId;
    this.primaryTag = this.object.primaryTag;
    this.recTokens = this.object.recTokens;
    this.isListed = this.object.isListed;
    this.recTokensLength = this.object.recTokens.size;
    this.recommendationRankings = new Map();
    this.minRankings = [];
  }

  recommendSnippets() {
    const recommendableSnippets = this.recommendableSnippets(
      this.object.languageId,
      this.object.primaryTag
    );

    recommendableSnippets.forEach(snippet => {
      // Skip if the snippet is the same as the current snippet
      if (snippet.id === this.id) return;
      // Skip if the snippet is the same in another language
      if (snippet.slugId === this.slugId) return;
      // Skip unless this snippet is listed or the object is unlisted
      if (!snippet.isListed && this.isListed) return;

      // Store the minimum ranking so far
      const minRanking = this.minRankings[0] || 0;

      // Performance optimization - if language score is 0 and the minimum
      // recommendation score is greater than the score limit without language,
      // then we can skip the rest of the calculations.
      const isSameLanguage = this.languageId === snippet.languageId;
      if (
        !isSameLanguage &&
        minRanking > settings.recommendations.scoreLimitWithoutLanguage
      )
        return;

      // Determine score for language:
      //  * Same language, as language = 100% of language score
      //  * Not same language = 0% of language score
      const languageScore = isSameLanguage
        ? settings.recommendations.fullLanguageScore
        : 0;

      // Performance optimization - if both language and primary tag scores are
      // 0 and the minimum recommendation score is greater than the score limit
      // without language and primary tag, then we can skip the rest of the
      // calculations.
      const primaryTagIndex = isSameLanguage
        ? snippet.tags.indexOf(this.primaryTag)
        : -1;

      if (
        !isSameLanguage &&
        primaryTagIndex === -1 &&
        minRanking >
          settings.recommendations.scoreLimitWithoutLanguageAndPrimaryTag
      )
        return;

      // Determine primary tag score:
      //  * Different language = 0% of tag score
      //  * Same primary tag = 100% of tag score
      //  * Contains primary tag, but not primary = 50% of tag score
      //  * Doesn't contain tag = 0% of language score
      const primaryTagScore =
        primaryTagIndex === -1
          ? 0
          : primaryTagIndex === 0
            ? settings.recommendations.fullPrimaryTagScore
            : settings.recommendations.halfPrimaryTagScore;

      // Determine recommendation token score:
      //  * Count found tokens and divide by total number of tokens
      const recTokenScore =
        ([...this.recTokens].reduce(
          (a, t) => (snippet.recTokens.has(t) ? a + 1 : a),
          0
        ) /
          this.recTokensLength) *
        settings.recommendations.recTokenScoreLimit;

      // Divide by the limit to get a value between 0 and 1
      const recommendationRanking =
        (languageScore + primaryTagScore + recTokenScore) /
        settings.recommendations.totalScoreLimit;

      // Performance optimization to minimize the number of times we have to
      // sort afterwards. As soon as the minimum amount of snippets has been
      // considered, we can start trimming off any snippets below the lowest
      // snippet's recommendation ranking.
      if (recommendationRanking > 0) {
        if (
          this.minRankings.length < settings.recommendations.recommendationCount
        ) {
          // First 4 snippets are always added
          this.minRankings.push(recommendationRanking);
        } else {
          // If the new ranking is lower than the lowest ranking, ignore it
          if (recommendationRanking < this.minRankings[0]) {
            return;
          } else {
            // Otherwise, replace the lowest ranking with the new ranking
            this.minRankings[0] = recommendationRanking;
          }
        }
        this.minRankings.sort((a, b) => a - b);

        this.recommendationRankings.set(snippet.id, [
          recommendationRanking,
          snippet.ranking,
          snippet,
        ]);
      }
    });

    return Array.from(this.recommendationRankings.values())
      .sort((a, b) => (a[0] === b[0] ? b[1] - a[1] : b[0] - a[0]))
      .slice(0, settings.recommendations.recommendationCount)
      .map(r => r[2]);
  }
}
